### ISSUE TITLE
Stuck at login

### ISSUE THREAD

**Issue by falematte** [martianhi/mastodon.el#149]
I am not able to connect to https://framapiaf.org/web/getting-started . Everything in emacs is stuck with the message "Contacting host: framapiaf.org:443"

**Comment by hdurer** [comment]
Oh dear.  We've had some other report in the past with similar symptoms for the `toot.love` server.
We never got to the bottom of that.

IIRC we tracked it down to something low-level in Emacs, i.e. not specific to mastodon.el, namely something simple as 
```
(let ((host "toot.love"))
  (with-temp-buffer
    (open-network-stream host (current-buffer) host 443 :type 'tls)))
```
was hanging.

Is that also true on your system for framapiaf.org?

Neither of these two hosts make this hang for me on my system...

**Comment by falematte** [comment]
Thank you for your answer. Could you please explain me how to troubleshoot properly? I have no idea on how to answer to your question...

**Comment by alexjgriffith** [comment]
I get the same issue when trying to log into  `https://framapiaf.org`. The http response for me indicates that that mastodon.el is not a valid client and the auth request is rejected. 

@falematte do you mind restarting emacs running `M-x mastodon` and copying the contents of the ` *http<your-server-name>*` buffer? Note that there is a space before the `*`.

**Comment by alexjgriffith** [comment]
```
﻿﻿﻿﻿﻿﻿﻿﻿HTTP/1.1 401 Unauthorized
Date: Wed, 28 Feb 2018 05:27:33 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Cache-Control: no-store
Pragma: no-cache
WWW-Authenticate: Bearer realm="Doorkeeper", error="invalid_client", error_description="L’authentification du client a échoué à cause d’un client inconnu, d’aucune authentification de client incluse ou d’une méthode d’authentification non prise en charge."
Vary: Accept-Encoding
Content-Encoding: gzip
X-Request-Id: ca57c667-4c52-4d60-96f7-f97653d04e51
X-Runtime: 0.055562

<hr>Sorry, but I do not know how to handle https://framapiaf.org//oauth/token authentication.  If you'd like to write it, please use M-x report-emacs-bug RET.<hr>
```

**Comment by hdurer** [comment]
`https://framapiaf.org//oauth/token authentication` ...  those two consecutive slashes look like a potential issue...

**Comment by hdurer** [comment]
Never mind...  I get the same response (but only single slash) when I try it after an `(setq mastodon-instance-url "https://framapiaf.org")`.

However, for me that seemed to be that my system had stale data from a previous connection to a different server.
We are not very good with switching between servers.  State is kept in `mastodon-client--token-file` which is specific the the server but we never ensure that the data is cleared/thrown away/ignored if the host has changed.

Anyway, after removing that file and resetting `mastodon-client--client-details` to nil, I get to the timeline.

Could it be that you had first called `M-x mastodon` and later changed the value of `mastodon-instance-url` to something else?

**Comment by alexjgriffith** [comment]
@hdurer your steps worked I can now see the timelines

Execute the following to set a new server and reset the credentials for your previous server.
``` emacs-lisp
(setq mastodon-instance-url "https://framapiaf.org")
(setq mastodon-client--client-details nil)
(setq mastodon-auth--token nil)
(when (file-exists-p mastodon-client--token-file)
 (dired-delete-file mastodon-client--token-file))
```

You can then run `M-x mastodon` like normal.

Perhaps this should be worked into a function somewhere?

@falematte do these steps work for you?

**Comment by falematte** [comment]
@alexjgriffith adding the lines that you suggested in my init file I got the following error:

``Symbol's value as variable is void: mastodon-client--token-file``

Removing the last two lines the error is gone but everything works. Thank you!

Anyway if I now press "f" or "b" I get "wrong type agument: numberp. "996..."

**Comment by alexjgriffith** [comment]
Yeah `mastodon-client--token-file` is not initialized until mastodon is run. If you run into the issue again you can execute the entire above block in `*scratch*` (after running `M-x mastodon`).

The numberp error was brought up in issue #150. It has been fixed in the latest develop branch.

## Installing Develop
If you want to use the develop branch you can try something like

Uninstall mastodon by removing the mastodon folder from your elpa file, which will be located somewhere like `~/emacs.d/elpa/`. If emacs is running restart it now.

Clone mastodon

``` sh
git clone https://github.com/jdenen/mastodon.el.git
```

Let emacs know where mastodon has been cloned to. Make sure to point to the `lisp` subfolder in the `mastodon.el` project.
``` emacs-lisp
(add-to-list 'load-path "/path/to/mastodon.el/lisp")
(require 'mastodon)
```

**Comment by alexjgriffith** [comment]
Hopefully closed by @hdurer PR b774f9e #154
